### ISSUE TITLE
[BR]: `fail2ban-client start` fails with Docker environments due to excessively large FD limits to iterate and close

### ISSUE THREAD

**Issue by polarathene** [fail2ban/fail2ban#3334]
**UPDATE:**

I've since investigated and discovered the cause is here:

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/server/server.py#L917-L935

- In environments like Docker, the FD limit may commonly be approx 1 million or 1 billion depending on the OS configuration (_at least with systemd service files setting `LimitNOFILE=` for `docker.service`_).
- Even though only a few standard FDs may be open, the excessive range will be iterated consuming heavy CPU. In my experience this was 8 minutes for `os.closerange()` and approx 70 minutes for the alternative `os.close()` loop.

A workaround when possible is to set a lower limit for the container with the `docker run` option `--ulimit`. Some users have reported not being able to use this due to lack of permissions required in their environment.

Various solutions have been suggested in the discussion below, but the easiest approach may be to set a sane maximum to close and log when that is exceeded to inform users to explicitly request a larger/unrestricted range if needed?

---

### Reproduction

This is taken from a later comment. With Docker installed this command should fail (_at least on linux hosts with systemd_):

```console
$ docker run --rm -it alpine ash -c 'apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && fail2ban-client start && ps aux | grep fail2ban'

# ...
2022-08-16 05:20:19,777 fail2ban                [11]: ERROR   Could not start server. Maybe an old socket file is still present. Try to remove /var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
```

While this will be successful (_`--ulimit` option sets a reduced limit that is possible to iterate through before the 30 second default timeout_):

```console
$ docker run --rm -it --ulimit 'nofile=1024' alpine ash -c 'apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && fail2ban-client start && ps aux | grep fail2ban'

# ...
Server ready
# ...
   14 root      0:00 {fail2ban-server} /usr/bin/python3 /usr/bin/fail2ban-server --async -b -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid --loglevel INFO --logtarget /var/log/fail2ban.log --syslogsocket auto
```

If one knows how much time is needed to be successful, raising the timeout limit also works: `fail2ban-client --timeout 600 start` (10 minutes). Otherwise when the command times out, the `fail2ban-server` process continues closing file descriptors and isn't killed from the timeout.

---

<details>
<summary>Original report</summary>

### Environment:

Running within a Docker container (Debian Bullseye for production, but have reproduced in Alpine). Docker daemon is running on a VM guest of Fedora 36 (hypervisor VMware  Workstation 16.3) on a Arch Linux host system (x86_64).

- Fail2Ban version: `fail2ban-0.11.2-r2`
- OS, including release name/version : Alpine Linux v3.16 (Docker container)
- [x] Fail2Ban installed via OS/distribution mechanisms
- [x] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:

Failure occurs when attempting to start `fail2ban` in a Docker container:

```console
$ fail2ban-client start

2022-08-16 04:16:46,049 fail2ban                [28]: ERROR   Could not start server. Maybe an old socket file is still present. Try to remove /var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
```

<details>
<summary>Log output (with `-vvv -x`)</summary>

```console
$ fail2ban-client -vvv -x start

 +  141 7FE0EAB16B48 fail2ban.configreader     INFO  Loading configs for fail2ban under /etc/fail2ban 
 +  143 7FE0EAB16B48 fail2ban.configreader     DEBUG Reading configs for fail2ban under /etc/fail2ban 
 +  144 7FE0EAB16B48 fail2ban.configreader     DEBUG Reading config files: /etc/fail2ban/fail2ban.conf
 +  145 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/fail2ban.conf']
 +  145 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Reading file: /etc/fail2ban/fail2ban.conf
 +  146 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/fail2ban.conf']
 +  148 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Shared file: /etc/fail2ban/fail2ban.conf
 +  149 7FE0EAB16B48 fail2ban                  INFO  Using socket file /var/run/fail2ban/fail2ban.sock
 +  149 7FE0EAB16B48 fail2ban                  INFO  Using pid file /var/run/fail2ban/fail2ban.pid, [INFO] logging to /var/log/fail2ban.log
 +  150 7FE0EAB16B48 fail2ban                  HEAVY  -- ['ping', 0.1] failed -- FileNotFoundError(2, 'No such file or directory')
 +  151 7FE0EAB16B48 fail2ban.configreader     INFO  Loading configs for jail under /etc/fail2ban 
 +  152 7FE0EAB16B48 fail2ban.configreader     DEBUG Reading configs for jail under /etc/fail2ban 
 +  154 7FE0EAB16B48 fail2ban.configreader     DEBUG Reading config files: /etc/fail2ban/jail.conf, /etc/fail2ban/jail.d/alpine-ssh.conf
 +  155 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/jail.conf']
 +  155 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Reading file: /etc/fail2ban/jail.conf
 +  169 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/paths-debian.conf']
 +  171 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Reading file: /etc/fail2ban/paths-debian.conf
 +  173 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/paths-common.conf']
 +  174 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Reading file: /etc/fail2ban/paths-common.conf
 +  188 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/paths-overrides.local']
 +  194 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Reading file: /etc/fail2ban/paths-overrides.local
 +  200 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/jail.d/alpine-ssh.conf']
 +  206 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Reading file: /etc/fail2ban/jail.d/alpine-ssh.conf
 +  207 7FE0EAB16B48 fail2ban.configparserinc  INFO    Loading files: ['/etc/fail2ban/paths-common.conf', '/etc/fail2ban/paths-debian.conf', '/etc/fail2ban/jail.conf', '/etc/fail2ban/jail.d/alpine-ssh.conf']
 +  207 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Shared file: /etc/fail2ban/paths-common.conf
 +  207 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Shared file: /etc/fail2ban/paths-debian.conf
 +  207 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Shared file: /etc/fail2ban/jail.conf
 +  208 7FE0EAB16B48 fail2ban.configparserinc  TRACE     Shared file: /etc/fail2ban/jail.d/alpine-ssh.conf
 +  249 7FE0EAB16B48 fail2ban                  DEBUG   async starting of server in 23, fork: True - 24
 +  252 7FE0EAB16B48 fail2ban                  HEAVY __waitOnServer: (True, 30)
 +  253 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.0009500980377197266
 +  267 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.014755487442016602
 +  287 7FE0EAB16B48 fail2ban                  DEBUG   async starting of server in 24, fork: True - 0
 +  290 7FE0EAB16B48 fail2ban                  DEBUG Starting '/usr/bin/fail2ban-server' with args ['fail2ban-server', '--async', '-b', '-s', '/var/run/fail2ban/fail2ban.sock', '-p', '/var/run/fail2ban/fail2ban.pid', '-x', '-vvv', '--loglevel', 'INFO', '--logtarget', '/var/log/fail2ban.log', '--syslogsocket', 'auto']
 +  295 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.04277944564819336
 +  348 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.09576845169067383
 +   69 7FBB6DACAB48 fail2ban                  INFO  Using socket file /var/run/fail2ban/fail2ban.sock
 +   73 7FBB6DACAB48 fail2ban                  INFO  Using pid file /var/run/fail2ban/fail2ban.pid, [INFO] logging to /var/log/fail2ban.log
 +   74 7FBB6DACAB48 fail2ban                  DEBUG   direct starting of server in 24, deamon: True
 +  457 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.2051231861114502
 +  151 7FBB6DACAB48 fail2ban.server           INFO  Starting in daemon mode
 +  153 7FBB6DACAB48 fail2ban                  DEBUG Exit with code 0
 +  658 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.4056081771850586
 + 1065 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 0.8127536773681641
 + 1567 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 1.3147380352020264

INFO   [ + 2073 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 1.8207433223724365
  + 2574 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 2.321540355682373
  + 3075 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 2.8223278522491455
  + 3581 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 3.3287596702575684
  + 4084 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 3.8317525386810303
  + 4587 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 4.3347508907318115
  + 5088 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 4.835473299026489
  + 5591 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 5.338402271270752
  + 6091 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 5.838769435882568
  + 6599 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 6.346776962280273
  + 7102 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 6.849779844284058
 + 7606 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 7.353386640548706
 + 8109 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 7.856667757034302
 + 8610 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 8.35775375366211
 + 9113 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 8.860753774642944
 + 9614 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 9.361480474472046
 +10118 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 9.865747928619385
 +10620 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 10.367749452590942
 +11127 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 10.874731302261353
 +11628 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 11.375521421432495
 +12131 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 11.878766298294067
  +12634 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 12.381748676300049
  +13140 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 12.887734413146973
  +13643 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 13.390751361846924
  +14152 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 13.89975881576538
  +14653 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 14.400509119033813
  +15156 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 14.903644323348999
  +15660 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 15.407699823379517
  +16161 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 15.908511638641357
  +16666 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 16.413676023483276
  +17173 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 16.920743465423584
 +17676 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 17.423735857009888
 +18177 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 17.92453360557556
 +18684 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 18.431763410568237
 +19186 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 18.933747053146362
 +19687 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 19.43473505973816
 +20195 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 19.94274401664734
 +20698 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 20.445739030838013
 +21202 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 20.94961166381836
 +21705 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 21.45273733139038
 +22213 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 21.96073269844055
  +22716 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 22.463761806488037
  +23219 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 22.96674084663391
  +23723 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 23.470768928527832
  +24233 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 23.980743169784546
  +24736 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 24.483736753463745
  +25239 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 24.9867525100708
  +25740 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 25.487753868103027
  +26244 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 25.991734981536865
  +26745 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 26.492750644683838
  +27248 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 26.995726823806763
 +27752 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 27.49959707260132
 +28265 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 28.01280903816223
 +28773 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 28.52074670791626
 +29276 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 29.023734092712402
 +29777 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 29.524730682373047
 +30282 7FE0EAB16B48 fail2ban                  HEAVY   wait-time: 30.029735803604126
 +30282 7FE0EAB16B48 fail2ban                  ERROR Failed to start server
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/fail2ban/client/fail2banclient.py", line 326, in __processStartStreamAfterWait
    if not self.__waitOnServer(): # pragma: no cover
  File "/usr/lib/python3.10/site-packages/fail2ban/client/fail2banclient.py", line 364, in __waitOnServer
    raise ServerExecutionException("Failed to start server")
fail2ban.client.fail2bancmdline.ServerExecutionException: Failed to start server
 +30284 7FE0EAB16B48 fail2ban                  ERROR Could not start server. Maybe an old socket file is still present. Try to remove /var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
 +30286 7FE0EAB16B48 fail2ban                  DEBUG Exit with code 255
```

</details>

The command stalls for 30 seconds when a timeout seems to abort.

**Quick Fix:** Setting a lower `ulimit -n` value (_normally 1024 for a linux OS, Docker containers instead set `ulimit -n` to over a billion by default_) works around the bug (_assuming the failure is due to attempting to perform a billion system calls to close file descriptors_).

#### Steps to reproduce

1. With Docker installed, an Alpine container can reproduce this on a Fedora 36 VM guest (_although I don't expect the host OS to matter much, at least for Linux hosts_).
2. `docker run --rm -it alpine ash` is sufficient, then install the `fail2ban` package `apk add fail2ban`.
3. The command will fail early due to the default jail being shipped for SSH. Disable it in the config or remove it `rm /etc/fail2ban/jail.d/alpine-ssh.conf`.
4. Run the command `fail2ban-client start`, it should fail with a timeout internally after 30 seconds, and output the error shown above.

If you were not able to reproduce for some reason, verify these container and host results:

```conse
# Docker container (alpine)
$ getconf -a | grep '^OPEN_MAX'

OPEN_MAX                           1073741816
# NOTE: `_POSIX_OPEN_MAX` on alpine is similar but set to 20,
# on Debian bullseye it's the same `1073741816` value.


# VM guest Fedora 36 (Docker host)
$ getconf -a | grep '^OPEN_MAX'

OPEN_MAX                           1024
# NOTE: `_POSIX_OPEN_MAX` is also 1024 in this case.

# NOTE: `ulimit -n` and `sysctl fs.nr_open` also outputs the same value,
# While `ulimit -n` will reflect an adjusted value,
# `sysctl fs.nr_open` will always be `1073741816`, even outside the container.
```

#### Expected behavior

Command should be successful:

```console
$ fail2ban-client start

Server ready
```

#### Observed behavior

I recently [debugged a similar problem with another package running in a Docker container](https://github.com/roehling/postsrsd/issues/122#issuecomment-1214641239), I believe it's the same issue here.

I am not sure where to look in `fail2ban` source code, or if it's an upstream dependency of `fail2ban` (possibility).

The problem is very much likely due to an attempt to iterate through all file descriptors and closing them while starting the server as a daemon, or some similar action.

AFAIK this is a common step to apply while starting a daemon service, with naive approach to detect the FD limit from the system and iterate through each with a `close()` call.

#### Any additional information

The cause (_if I'm correct_) is a fault of Dockers.

When running a command via the root user (_common as it's the default implicitly_) within the Docker container, `ulimit -n` and similar methods will report a value exceeding 1 billion. On actual host systems (_and when running commands as a different user in the Docker container_), there is a more sane limit of 1024 usually reported (_less on macOS AFAIK_).

Docker (_or rather `containerd` I believe_) needs this high limit apparently to support many containers running at the same time, but the containers themselves rarely need it AFAIK. A workaround is to either set the Docker daemon config (_if possible to configure, eg not permitted in Github CI?_) or per container with the `--ulimit 'nofile=1024'` option.

Within the running container for reproduction, you can also adjust this limit internally once (_it'll fail if you try to adjust it again later_) with `ulimit -n 1024`. Now the same command will run successfully as expected.

The failure has nothing to do with what the error output is suggesting. In the other project that had this bug, using it within Docker resulted in approximately 10 minute startup time (vs <1 second). It was a million times slower starting up due to this issue.

In the production container (Debian Bullseye), the issue is reproducible locally while I develop, but doesn't appear to be caught by our Github CI tests. Probably due to Github configuring the Docker daemon to use a lower limit, or potentially the Ubuntu 20.04 environment used, ~~I have not investigated that yet~~ (_it is due to systemd services configured in `/usr/lib/systemd/system/` set fd limits via `LimitNOFILE=`. Fedora sets `docker.service` and `containerd.service` to `infinity`, while Manjaro sets `docker.service` to `1048576`_).

The issue itself is known upstream in Docker for several years now, with multiple python projects reported as hitting this issue when used in Docker containers.

---

## Possible fix

If you can identify where this issue is occurring (_in `fail2ban` or upstream dependency, or Python itself_), a fix can be:
- To use `close_range()` if available instead of the `close()` syscall.
- As that may not be as widely available, a solution RedHat took for `rpm` that also had this problem was to iterate through entries in `/proc/self/fd`.
- While another project iterated a fixed 1024 value and would continue until the first error from `close()` (FD doesn't exist), assuming that no other FDs would exist beyond that.

Meanwhile, a valid workaround is to add the [`--ulimit 'nofile=1024'` option](https://docs.docker.com/engine/reference/commandline/run/#set-ulimits-in-container---ulimit) on their container config.

</details>

**Comment by polarathene** [comment]
One liner reproduction commands (assuming Docker is installed). Give the failure case 30 seconds to timeout before the error shows. I've used `# ...` to omit output that isn't of interest.

---

**Failure (default `ulimit -n`):**
```console
$ docker run --rm -it alpine ash -c 'apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && fail2ban-client start && ps aux | grep fail2ban'

# ...
2022-08-16 05:20:19,777 fail2ban                [11]: ERROR   Could not start server. Maybe an old socket file is still present. Try to remove /var/run/fail2ban/fail2ban.sock. If you used fail2ban-client to start the server, adding the -x option will do it
```

**Success (sane `ulimit -n`):**
```console
docker run --rm -it --ulimit 'nofile=1024' alpine ash -c 'apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && fail2ban-client start && ps aux | grep fail2ban'

# ...
Server ready
# ...
   14 root      0:00 {fail2ban-server} /usr/bin/python3 /usr/bin/fail2ban-server --async -b -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid --loglevel INFO --logtarget /var/log/fail2ban.log --syslogsocket auto
```

**Comment by sebres** [comment]
> AFAIK this is a common step to apply while starting a daemon service, with naive approach to detect the FD limit from the system and iterate through each with a close() call.

I don't think the descriptors limit or something like this causing the issue.
Looks like it is something wrong with waiting for the forked process or detection of its start.
The first question does this forked fail2ban server process exists at all? I mean do you see a second python process with fail2ban?
Also you can try to start fail2ban-server directly, without fork, for instance in foreground like our systemd-unit doing it:
```
?sudo? fail2ban-server -xf --logtarget=stdout start
```
Do you see it starting in foreground?

**Comment by polarathene** [comment]
> I don't think the descriptors limit or something like this causing the issue.
> Looks like it is something wrong with waiting for the forked process or detection of its start.

Then why is it reproducible with the two one-liner snippets?

If you cannot use those to reproduce the failure, you can use `--ulimit 'nofile=infinity'` option with `docker run` which should raise it to the billion limit that I get on my end.

Setting a considerably lower limit to what is usually experienced outside the container does resolve the problem too. I'd say that's a pretty good indication that it's got something to do with it when that's the only thing that needs to change.

---

I just checked, and increasing the limit steadily increases the time taken for `fail2ban-client start` to run. This takes 5 seconds for me (approx 10 million):

```console
$ docker run --rm -it --ulimit 'nofile=10240000' alpine ash -c 'apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && time fail2ban-client start && ps aux | grep fail2ban'

# ...
Server ready
real    0m 5.66s
user    0m 0.13s
sys     0m 0.03s
    1 root      0:00 ash -c apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && time fail2ban-client start && ps aux | grep fail2ban
   15 root      0:04 {fail2ban-server} /usr/bin/python3 /usr/bin/fail2ban-server --async -b -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid --loglevel INFO --logtarget /var/log/fail2ban.log --syslogsocket auto
   19 root      0:00 grep fail2ban
```

> The first question does this forked fail2ban server process exists at all? I mean do you see a second python process with fail2ban?

Based on the above output, I guess it does? As I mentioned it appeared to exit at 30 seconds as a timeout, which wouldn't happen if the limits were reduced (_as shown above, the higher it goes the longer the time to complete the `fail2ban-client start` command takes_).

---

> Also you can try to start fail2ban-server directly, without fork, for instance in foreground like our systemd-unit doing it:
> Do you see it starting in foreground?

```bash
docker run --rm -it --ulimit 'nofile=1024000000' alpine ash -c 'apk add fail2ban && rm /etc/fail2ban/jail.d/alpine-ssh.conf && time fail2ban-server -xf --logtarget=stdout start && ps aux | grep fail2ban'
```

With a limit over a billion it has no slow down, no trouble getting to "Server ready".

**Comment by sebres** [comment]
OK, 188cdda0f660e29dcc27c7bf31f64178f8755ee1 indeed shows that tests `testClientStartBackground*` pretty reproducible [fail](https://github.com/fail2ban/fail2ban/runs/7856089449?check_suite_focus=true#step:8:535) with large `nofile` limit (I guess exceeding internal timeout). I'll check what exactly can cause it.

Just if the issue is really like slow [subprocess.Popen](https://bugzilla.redhat.com/show_bug.cgi?id=1820813), it can be a problem for entire fail2ban functionality, since fail2ban uses it to execute action commands: https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/server/utils.py#L194

**Comment by polarathene** [comment]
## Debugging journey

### Looking into `subprocess.Popen()` and `close_fds=False` parameter

I tried to use the `close_fds=False` parameter for `subprocess.Popen()` at the location you referenced, but it made no difference to the behaviour.

From the python [`subprocess` docs](https://docs.python.org/3/library/subprocess.html#popen-constructor):

> If _close_fds_ is true, all file descriptors except `0`, `1` and `2` will be closed before the child process is executed. Otherwise when _close_fds_ is false, file descriptors obey their inheritable flag as described in [Inheritance of File Descriptors](https://docs.python.org/3/library/os.html#fd-inheritance).

**UPDATE:** I suppose that's a non-issue due to the behaviour I discovered in `fail2ban-server` closing all the FDs upfront?

I tried to look through the [Python source for the `subprocess.Popen()`](https://github.com/python/cpython/blob/230b630a767a457c6e7e4d797548a172c64ae735/Lib/subprocess.py#L1685) but could not track down how `close_fds` parameter was being handled.

---

### Increasing the timeout option allows the server to successfully start with the client not bailing early

I used the `--timeout` option of `fail2ban-client` which overrides the default timeout configured (30 secs):

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2bancmdline.py#L47

With a value of 10 minutes (600), the command completes successfully at about the 8 minute mark (with the billion FD limit).

<details>
<summary>Earlier attempt - wrong direction</summary>

Tracking the progress further, the command was not completing this processing function, only reaching this point:

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L94

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/csocket.py#L43

Which turns out to be due to an exception triggered (_if condition is false, the exception just returns `False`_):

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L109-L115

**UPDATE:** I probably should have monitored what the actual command was being processed... This was due to `ping` being called internally for the `start` command here:

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L160-L163

</details>

---

### Timeout location found

The `fail2ban-client start` command does not pass this point:

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L191

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L323-L326

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L344-L353

It's waiting on the server, until the timeout occurs to stop here.

#### Timeout for client exits, but server continues to run (can create multiple until initial server is up)

Despite the timeout failure, the `fail2ban-server` process continues doing it's thing, so that once it is finally up, a `fail2ban-client start` command will by that point say a server is already running..

But until that happens, multiple `fail2ban-server` processes can be started with each `fail2ban-client start` command run.

---

## Actual problem:

So I dug into it further and eventually stumbled upon this (_after a long-winded process trying to navigate the path / flow of `fail2ban-client start` and where the hang was starting_)

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/server/server.py#L909-L915

After that snippet defines the value of `maxfd`, it then chooses one of these two methods to call (_in my case it was the `os.closerange()` call_):

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/server/server.py#L917-L935

**That's where the stall was**, as suspected originally.

I wasn't aware that the [`os.closerange()`](https://docs.python.org/3/library/os.html#os.closerange) call would be this slow, I had been given the impression it'd be faster (_the docs imply that, but perhaps usage in Docker affects that somehow, or just the massive range in itself_).

I tried to log after the `os.closerange()` call in the condition (_and afterwards with the remaining logic_), but that output never appeared on the terminal (_perhaps it eventually raises an exception_)?

**EDIT:** I modified the condition to run the loop of `os.close()` calls instead, and it is much slower. The 10 minute timeout was reached, but the process time from `ps aux | grep fail2ban-server` showed 60 minutes uptime when I went to check, with the CPU usage remaining high.

### Solutions

Iterating through `/proc/self/fd` for actual FDs in use to close would probably be a far better and more efficient solution? This is what RedHat chose to do for the issue with `rpm`.

Others chose a sane limit (eg: 1024) and exit as soon as an `os.close()` fails afterwards (_might not be entirely correct, but AFAIK it's highly unlikely for a random FD number to be in use explicitly at such a high range_).

I'm not sure what scenarios would have `fail2ban` daemonize with inheriting such a large FD range that's actually due to software using it? `ls /proc/self/fd` for example in my case is just 4 (FDs: 0, 1, 2 ,3). It may be more common that users experience this stall (_and the heavy CPU usage that continues after timeout_), than there is users with excessive FD ranges not being closed (_the kind of user that is more than likely capable of filing a bug report as they know their special case_). A debug log could also indicate if opt-ing to not iterate through an excessive range to close.

### Some info about FD limits history and software usage if helpful

With macOS the limit is 256 apparently, which is what your code falls back to if it cannot determine `maxfd`. I don't have access to macOS to know if that has since changed.

#### Linux kernel defaults and a case where explicit FDs instead of incremental FD numbers may be used

I am aware of advice around the internet to use / create the FD 200 for locking with `flock()`, and the Linux kernel defaults to soft limit of 1024 (_due to compatibility with `select()`, which I recall were referenced in the Python docs/source_) but raised the hard limit to 4096 in 2011 (_apps have to opt-in to a higher soft limit, but cannot exceed this hard limit_).

#### Systemd enters the mix (plus server software that advises larger limits for config)

Systemd systems with v240 (Dec 2018) then came along with overrides for systemd managed services with `LimitNOFILE=` setting to raise this so services like Kafka (advises 64k to 128k limit) can be better supported (_this is the opt-in raising of soft limit AFAIK_). v240 introduced a global default for services that kept the soft limit of 1024 for compatibility, but raise the soft limit to 512k.

#### Docker specific

Depending on the distribution, I've found this varies for Docker. `containerd.service` on Manjaro and Fedora are set to `infinity` (the 1 billion value), while `docker.service` is also `infinity` on Fedora, but about 1 million on Manjaro.

Still fairly excessive, but defaults to support most software needs in containers I guess, but some software that daemonizes balks at the much higher limits due to all the syscall overhead (_mostly iterating through FDs to close that aren't open and just returning errors, just in-case_).

---

## Off-topic (_deprecated use of Thread `setDaemon()`_)

Note, while my troubleshooting went through the AsyncServer route only, I did notice that this snippet has a deprecated method `setDaemon()`:

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L193-L201

https://docs.python.org/3/library/threading.html#threading.Thread.setDaemon

> Deprecated getter/setter API for [daemon](https://docs.python.org/3/library/threading.html#threading.Thread.daemon); use it directly as a property instead.
> 
> Deprecated since version 3.10.

I am not that familiar with Python, but based on the [docs for Threading](https://docs.python.org/3/library/threading.html#threading.Thread) there is an init `daemon` parameter that can be set. I assume you're setting that here:

https://github.com/fail2ban/fail2ban/blob/ff7fe572bf90010ee40c9d9966e7f5ef3ee4a145/fail2ban/client/fail2banclient.py#L217-L222

**Comment by polarathene** [comment]
The original code for this was [introduced to `fail2ban` in 2006, back when using SVN](https://github.com/fail2ban/fail2ban/commit/704a0b834f8a55240e2db1e4fc15bb3d434c2489), so no helpful context.

A comment that still exists references a recipe for the code introduced, that [URL still works (but redirects)](https://code.activestate.com/recipes/278731/). There was a total of 6 revisions with the last one made 16 years ago (roughly around the time this was introduced), looking through the latest revision (only one available) and the comment history, it would seem the contribution commit referenced an earlier revision. 

You may want to update that code based on the newer recipe revision (_or whatever modern practices / advice for doing this exist today with Python_). Note that it specifically sets a higher maximum of 1024 and enforces this when the limit checked is too high:

```python
   # Close all open file descriptors.  This prevents the child from keeping
   # open any file descriptors inherited from the parent.  There is a variety
   # of methods to accomplish this task.  Three are listed below.
   #
   # Try the system configuration variable, SC_OPEN_MAX, to obtain the maximum
   # number of open file descriptors to close.  If it doesn't exists, use
   # the default value (configurable).
   #
   # try:
   #    maxfd = os.sysconf("SC_OPEN_MAX")
   # except (AttributeError, ValueError):
   #    maxfd = MAXFD
   #
   # OR
   #
   # if (os.sysconf_names.has_key("SC_OPEN_MAX")):
   #    maxfd = os.sysconf("SC_OPEN_MAX")
   # else:
   #    maxfd = MAXFD
   #
   # OR
   #
   # Use the getrlimit method to retrieve the maximum file descriptor number
   # that can be opened by this process.  If there is not limit on the
   # resource, use the default value.
   #
   import resource		# Resource usage information.
   maxfd = resource.getrlimit(resource.RLIMIT_NOFILE)[1]
   if (maxfd == resource.RLIM_INFINITY):
      maxfd = MAXFD
```

The [wikipedia article on steps required to daemonize a process](https://en.wikipedia.org/wiki/Daemon_%28computing%29#Unix-like_systems) also mentions closing all FDs, but that this can also be managed by service managers (such as systemd). So this might be code that only runs in environments like Docker I guess?

**Comment by sebres** [comment]
Thx for details of your journey...

Strange is that I cannot reproduce it on my Debian (bullseye), I see only few hundert extra milliseconds:
<details><summary>tests with python 2.7</summary>

```diff
- $ sudo sh -c "ulimit -n 1048576 && sudo -u $LOGNAME bin/fail2ban-testcases --fast -vvv testClientStartBackground"
+ $ sudo sh -c "ulimit -n    1048 && sudo -u $LOGNAME bin/fail2ban-testcases --fast -vvv testClientStartBackground"
  Fail2ban 0.10.6 test suite. Python 2.7.18 (default, Jul 14 2021, 08:11:37) [GCC 10.2.1 20210110]. Please wait...
- testClientStartBackgroundCall (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.462s -- ok
+ testClientStartBackgroundCall (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.086s -- ok
- testClientStartBackgroundInside (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.237s -- ok
+ testClientStartBackgroundInside (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.049s -- ok
- Ran 2 tests in 0.699s
+ Ran 2 tests in 0.135s
```
</details>
<details><summary>tests with python 3.9</summary>

```diff
- $ sudo sh -c "ulimit -n 1048576 && sudo -u $LOGNAME bin/fail2ban-testcases --fast -vvv testClientStartBackground"
+ $ sudo sh -c "ulimit -n    1048 && sudo -u $LOGNAME bin/fail2ban-testcases --fast -vvv testClientStartBackground"
  Fail2ban 0.10.6 test suite. Python 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]. Please wait...
- testClientStartBackgroundCall (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.507s -- ok
+ testClientStartBackgroundCall (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.145s -- ok
- testClientStartBackgroundInside (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.260s -- ok
+ testClientStartBackgroundInside (fail2ban.tests.fail2banclienttestcase.Fail2banClientTest) ...  0.075s -- ok
- Ran 2 tests in 0.767s
+ Ran 2 tests in 0.221s
```
</details>

However the execution times are definitely longer with large limit, it doesn't cause timeout and doesn't show the increase of demonization time up to 1000x (it shows not more than 5x "deterioration").
I cannot set the soft open file limit larger than 1048576 (nor even unlimited), but it is the same value I used to test in CI in 188cdda0f660e29dcc27c7bf31f64178f8755ee1.
Weird.

> Looking into subprocess.Popen() and close_fds=False parameter
I tried to use the close_fds=False parameter for subprocess.Popen() at the location you referenced, but it made no difference to the behaviour.

The location I referenced used in actions command, it has nothing to do with a daemonization process by forking.
I wanted just say that we could get still another performance issue with large open file limit.

**Comment by polarathene** [comment]
> I cannot set the soft open file limit larger than 1048576 (nor even unlimited)

A similar experience was shared by a user running Windows 10 with a Debian 11 VM guest, that ran the same AlpineLinux Docker container. They had the 1 million limit you mention and were unable to alter it for the container.

Your response wasn't clear to me if you're running that locally on a Debian 11 host, and avoiding the AlpineLinux Docker container entirely? (_Runs Python 3.10, which may contribute to it as it looks like they've made changes there and it was announced in the 3.10 changelog IIRC_)

On the host, if you run `fail2ban-client start` in a shell, it'll likely inherit the defaults from the kernel (1024 soft limit, 4096 hard limit) or whatever you have in `/etc/security/limits.conf` as override if any AFAIK. Whilst systemd will apply global default of 1024 soft and 512k hard limit for services IIRC, which individually may set their own via `LimitNOFILE=` in their `.service` file (_Docker is usually set to the 1 million value you cited, or inifinity in the case of Fedora_).

With the limit of `1048576` it only takes 1 second to run for me. This is the limit my Manjaro install would usually apply by default to the Docker container, but I am able to raise it to `1073741816` and reproduce the same delay of Fedora at approx 8 minutes. My hardware is a bit old (_Intel i5-6500 4 cores 4 threads, with plenty of other stuff running_) which may increase the impact.

By comparison, at `1024` it completes in a mere 33ms, so it might just be due to you not being able to raise the limit higher to reproduce.

**Comment by sebres** [comment]
It is a physical host in my case, not a VM...
I'm not in mode to test it with Docker (or even other containers that may do something wrong also), so I just used maximal limit I was able to set directly. 
I just wanted to hint that the issue, which is pretty visible in CI's containers, is completely not reproducible on my physical linux host **with the same value of limit**.

Anyway, 535a982dccfb5e05f8fa0f3965e5bf454716179d seems to fix it now (CI tests pass now, CI fails in further steps due to another issue probably related to previous sudo testing).
Test and review are welcome.

**Comment by polarathene** [comment]
I have taken a random example of using `subprocess.popen()`:

```python
from subprocess import Popen, PIPE

process = Popen(['cat', '/tmp/data.txt'], stdout=PIPE, stderr=PIPE)

stdout, stderr = process.communicate()

print(stdout)
```

Running that worked without any delay in the same Docker container. So it should only be an issue with the daemonization logic.

---

> It is a physical host in my case, not a VM...

I understand, I've tested on Manjaro as a host, but only via the Docker container. What does `ulimit -Hn` output for you? Is it 512k? Have you verified that `maxfd = os.sysconf("SC_OPEN_MAX")` in `server.py` is using the limits you set? It's likely the case since 1 million still seems to be processed fairly quickly.

**Comment by sebres** [comment]
> What does ulimit -Hn output for you? Is it 512k? 

32K

> Have you verified that maxfd = os.sysconf("SC_OPEN_MAX") in server.py is using the limits you set?

Sure, it is 1048576.
Although I need a hundredfold (100M) for noticeable delay.
But I guess even by linear growth of execution time it would need over 6 minutes for max signed int32 value (0x7fffffff) on my bare-metal system.
<details><summary>PoC</summary>

```
$ time python -c 'import os; os.closerange(3, 1000000)'
real    0m0.195s
user    0m0.104s
sys     0m0.091s
$ time python -c 'import os; os.closerange(3, 10000000)'
real    0m1.811s
user    0m1.239s
sys     0m0.571s
$ time python -c 'import os; os.closerange(3, 100000000)'
real    0m17.952s
user    0m12.144s
sys     0m5.808s
```
</details>

Anyway, you could try to apply my fix 535a982dccfb5e05f8fa0f3965e5bf454716179d and test it in your docker container...

**Comment by sebres** [comment]
Thus close (merged in all main branches - 0.10, 0.11, 1.0).
Thanks again for the contribution!

**Comment by polarathene** [comment]
Your solution works great! :grinning: (_although I've only tested time to run `fail2ban-client start` itself and `fail2ban-client stop` afterwards_)

Thanks for the quick resolution :+1: 

---

## For reference (Python internal FD closing logic)

If anyone lands here and is interested, I looked into what Python was doing behind the scenes and figured it might be helpful to others.

I tried the `subprocess.popen()` example on Python versions 3.5, 3.9, 3.10 and 3.11 (unreleased) and noticed no significant perf difference between versions despite the high limit. Likewise with `time python -c 'import os; os.closerange(3, 100000000)'`. 

Oct 2020 (_presumably for Python 3.10 and up_) [introduced an internal API `_Py_closerange()`](https://github.com/python/cpython/pull/19754) for both `os.closerange()` and `subprocess.popen()` to leverage for closing FDs AFAIK intended to later support the `close_range()` syscall which got [reworked in May 2022](https://github.com/python/cpython/pull/92303) to take priority over other strategies.

- [`os.closerange()`](https://github.com/python/cpython/blob/586fc02be5b3e103bfddd49654034a898a8d6dfc/Modules/posixmodule.c#L9282-L9300) still calls `_Py_closerange()` today (_which had it's code moved to [`fileutils.c`](https://github.com/python/cpython/blob/586fc02be5b3e103bfddd49654034a898a8d6dfc/Python/fileutils.c#L2611-L2649)_).  It prefers the `close_from()` syscall, then tries `closefrom()` syscall (_only if closing up to `sysconf(_SC_OPEN_MAX)`_), I'm not familiar with FDWalk support on linux, the last fallback is iterating the entire range with `close()` syscall.
- `subprocess.popen()` AFAIK internally calls [this `child_exec()` method closing FDs here](https://github.com/python/cpython/blob/586fc02be5b3e103bfddd49654034a898a8d6dfc/Modules/_posixsubprocess.c#L654-L658), which defers to [this `_close_open_fds()`method](https://github.com/python/cpython/blob/586fc02be5b3e103bfddd49654034a898a8d6dfc/Modules/_posixsubprocess.c#L422-L433).
- `_close_open_fds()` then will attempt to use `close_range()` syscall if available, otherwise for Linux AFAIK set the fallback method to [`_close_open_fds_safe()`](https://github.com/python/cpython/blob/586fc02be5b3e103bfddd49654034a898a8d6dfc/Modules/_posixsubprocess.c#L250-L326). That method appears to use `procfs` kernel feature to get a list of open FDs to close, otherwise if that approach is unavailable it resorts to using the `_brute_force_closer()` approach iterating through the entire range with `close()` syscall (_due to setting the `end_fd` parameter to `_close_range_except()` as `-1`, which calls [`safe_get_max_fd()`](https://github.com/python/cpython/blob/586fc02be5b3e103bfddd49654034a898a8d6dfc/Modules/_posixsubprocess.c#L185-L210)_)
-  `safe_get_max_fd()` appears to query the max limit (_with `sysconf(_SC_OPEN_MAX)`_) and if `-1` (_infinity or undefined?_) chooses `256`.

So it appears `os.closerange()` does not provide a `/proc/self/fd` closer approach like `subprocess.popen()` appears to leverage. While if available on Linux (kernel 5.9+) they both would prefer `close_range()` syscall first, which if it's [as fast as implied](https://github.com/python/cpython/issues/92301#issue-1225267154) would seem good.

I could not reproduce those linked `close_range()` results with the Python 3.11-rc docker images (_performance was same throughout versions tested_), so perhaps `close_range()` isn't being used in my environment for some reason.. likewise `os.closerange(3, os.sysconf("SC_OPEN_MAX"))` did not seem to trigger the `closefrom()` syscall.
